src/lib.rs

```diff
@@ -29,3 +31,162 @@ pub mod code_table_type {
+                char::from(*self).fmt(f)
+            }
+        }
+    };
+    ($Cp:ident(Complete, $ENCODING_TABLE:ident, $DECODING_TABLE:ident)) => {
+        cp_impl! { $Cp(Common, $ENCODING_TABLE, $DECODING_TABLE) }
+
+        impl CompleteCp for $Cp {}
+
+        impl From<u8> for $Cp {
+            fn from(value: u8) -> Self {
+                Self(value)
+            }
+        }
+
+        impl From<$Cp> for char {
+            fn from(value: $Cp) -> Self {
+                if value.0 < 128 {
+                    value.0 as char
+                } else {
+                    code_table::$DECODING_TABLE[usize::from(value.0 - 128)]
+                }
+            }
+        }
+    };
+    ($Cp:ident(Incomplete, $ENCODING_TABLE:ident, $DECODING_TABLE:ident)) => {
+        cp_impl! { $Cp(Common, $ENCODING_TABLE, $DECODING_TABLE) }
+
+        impl TryFrom<u8> for $Cp {
+            type Error = TryFromU8Error;
+
+            fn try_from(value: u8) -> Result<Self, Self::Error> {
+                if value < 128 || code_table::$DECODING_TABLE[usize::from(value - 128)].is_some() {
+                    Ok(Self(value))
+                } else {
+                    Err(TryFromU8Error)
+                }
+            }
+        }
+
+        impl From<$Cp> for char {
+            fn from(value: $Cp) -> Self {
+                if value.0 < 128 {
+                    value.0 as char
+                } else {
+                    code_table::$DECODING_TABLE[usize::from(value.0 - 128)].unwrap()
+                }
+            }
+        }
+    };
+    ($($Cp:ident($Type:ident, $ENCODING_TABLE:ident, $DECODING_TABLE:ident),)*) => {
+        $(cp_impl! { $Cp($Type, $ENCODING_TABLE, $DECODING_TABLE) })*
+    };
+}
+
+cp_impl! {
+    Cp437(Complete, ENCODING_TABLE_CP437, DECODING_TABLE_CP437),
+    Cp720(Complete, ENCODING_TABLE_CP720, DECODING_TABLE_CP720),
+    Cp737(Complete, ENCODING_TABLE_CP737, DECODING_TABLE_CP737),
+    Cp775(Complete, ENCODING_TABLE_CP775, DECODING_TABLE_CP775),
+    Cp850(Complete, ENCODING_TABLE_CP850, DECODING_TABLE_CP850),
+    Cp852(Complete, ENCODING_TABLE_CP852, DECODING_TABLE_CP852),
+    Cp855(Complete, ENCODING_TABLE_CP855, DECODING_TABLE_CP855),
+    Cp857(Incomplete, ENCODING_TABLE_CP857, DECODING_TABLE_CP857),
+    Cp858(Complete, ENCODING_TABLE_CP858, DECODING_TABLE_CP858),
+    Cp860(Complete, ENCODING_TABLE_CP860, DECODING_TABLE_CP860),
+    Cp861(Complete, ENCODING_TABLE_CP861, DECODING_TABLE_CP861),
+    Cp862(Complete, ENCODING_TABLE_CP862, DECODING_TABLE_CP862),
+    Cp863(Complete, ENCODING_TABLE_CP863, DECODING_TABLE_CP863),
+    Cp864(Incomplete, ENCODING_TABLE_CP864, DECODING_TABLE_CP864),
+    Cp865(Complete, ENCODING_TABLE_CP865, DECODING_TABLE_CP865),
+    Cp866(Complete, ENCODING_TABLE_CP866, DECODING_TABLE_CP866),
+    Cp869(Complete, ENCODING_TABLE_CP869, DECODING_TABLE_CP869),
+    Cp874(Incomplete, ENCODING_TABLE_CP874, DECODING_TABLE_CP874),
```

Comment:
I wouldn't like to maintain these `Complete`/`Incomplete` written by hand if possible.
`build.rs` automatically generates `ENCODING_TABLE_*` and `DECODING_TABLE_*`.

Reply:
Hmm, so we might be able to move this macro call into `build.rs`. Is that what you had in mind?

Reply:
> Is that what you had in mind?

Right. `write_*` functions are for code generation and you have only to modify (some of) them and maybe `generate_tables`.
I wonder if it's possible.